### ISSUE TITLE
MachSuite Tracker

### ISSUE THREAD

**Issue by sa2257** [cucapra/seashell#45]
This is to keep track of the status of MachSuites apps implementation. Subsequent posts will track each app.

Some translations:
- C = "Something Compiles to Fuse"
- T = "The Fuse program was tested with the harness"
- R = "The Fuse program is a straightforward 'translation' from the original C"

| App | C | T | R | Notes |
| - | - | - | - | - |
| gemm ncubed | ✅| | Yes | Simple kernel
| stencil2d | |
| gemm block | |
| stencil3d | |
| fft | ✅| | Yes |
| kmp | ✅| | Yes |
| radix sort | ✅| | Yes |
| merge sort | | | |
| spmv | ✅| | Yes |
| bfs | ✅| | Yes |
| md | ✅| | Yes |
| backprop | |
| nw | ✅| | Yes |
| aes | |
| viterbi | |

**Comment by rachitnigam** [comment]
Use #23

**Comment by sa2257** [comment]
**gemm ncubed**

- [ ] Remove library inclusion? It is needed by the host file too. So the header file seems a better place for it.
- [ ] Local float variables?
- [ ] reduction operation on `temp`
- [ ] should get an error for multi-write (multi-write due to unrolling) [multiwrite example](https://github.com/cucapra/seashell/blob/master/examples/errors/multiwrite.sea)

**Comment by sampsyo** [comment]
Hi! I don't have a strong feeling about whether MachSuite goes in #23 or its own issue (here). But either way, let's list _all_ the benchmarks. This can help you prioritize the ones to work on next and keep a global view of how far along you are.

For the list of "gaps" indicating what we need before we can fully implement the benchmark, let's try to be as specific as possible and, when it's available, link to the existing issue that tracks that problem. For example, does "Local float variables?" mean that it's impossible to declare local variables with type `float`? If so, that probably deserves its own issue (or at least a full explanation here).

**Comment by rachitnigam** [comment]
Yeah, I'd prefer specific issues that block benchmarks from being written.

**Comment by sa2257** [comment]
Hi, so Rachit prefers to have specific issues listed in #23. I'll repurpose this issue to keep track of MachSuite benchmark implementation.

**Comment by sa2257** [comment]
***Language features needed for MachSuite apps***

**gemm/ncubed** - nested loops and reductions

**gemm/blocked** - _gemm/ncubed_ + tiling
tiling needs index addition to access to arrays, so views are needed

**fft/strided** - _gemm/ncubed_ + exponential + serialization

**stencil/stencil2d** - _gemm/ncubed_ + index addition and mismatching matmul (filtering)
views are needed to achieve complex array accesses and index addition

**stencil/stencil3d** - _stencil/stencil2d_

**spmv/ellpack** - _gemm/ncubed_ + indirection

**spmv/crs** - _spmv/ellpack_ + index addition

**kmp** - _stencil/stencil2d_ + while loops and if conditions

**fft/transpose** - _fft/strided_ + LUTs

**sort/merge** - _stencil/stencil2d_ + if conditions and some method of serialization
I think this can be reasonably translated as well, if we were to do an iterative mergesort. I'm unclear on what recursion would mean in terms of writing a Seashell program (i.e. how that would be mapped to hardware is something I'm confused about.)

**sort/radix** - _sort/merge_ + shifters
Bitwise operation support would be helpful for translating this; they use some shifting; I think this could be translated reasonably well though.

**bfs/bulk** - _sort/merge_
need a good way to represent nodes of a graph; maybe something like a `struct`? I suppose we can also use an adjacency matrix or something. But it seems like it would be generally helpful to have some kind of "and"-type

**bfs/queue** - _bfs/bulk_ + queues

***Next phase of apps***
*  md
*  backprop
*  nw
*  aes
*  viterbi

**Comment by sa2257** [comment]
***Optimizations Needed***
**gemm/ncubed** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**gemm/blocked** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll with tiling, pipeline
**stencil/stencil2d** -  resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**stencil/stencil3d** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**spmv/ellpack** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**spmv/crs** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**kmp** - memory allocation, array partitioning, loop unroll, pipeline
**bfs/bulk** - memory allocation, array partitioning, loop unroll, pipeline
**bfs/queue** - memory allocation, array partitioning, loop unroll, pipeline
**fft/strided** - resource allocation (multiplier), memory allocation, array partitioning, loop unroll, pipeline
**fft/transpose** - allocation (multiplier and adder), memory allocation, array partitioning, loop unroll, pipeline
**sort/merge** - memory allocation, array partitioning, loop unroll, pipeline
**sort/radix** - memory allocation, array partitioning, loop unroll, pipeline

**Comment by sampsyo** [comment]
Cool! Maybe a big table would be a good way to represent this information? The rows could be benchmarks, and the columns could be features. Then the cell could contain a checkmark if the benchmark needs that feature. (There could be two groups of columns: expressiveness/features and optimizations.) This way, we could easily see what will affect the most benchmarks.

Some things where I could use expanded definitions:

- "Memory allocation": Do the benchmarks call `malloc`? If so, do they really need to do this, or is it just a convenience?
- "Resource allocation (multiplier)": What does this look like? What's the effect on the generated hardware?
- "Serialization": Is this like data structure serialization, or like preventing parallelism?
- "LUTs": What do these look like in the code? (Presumably you don't mean FPGA LUTs?)
- "mismatching matmul (filtering) views": This could use a little more explanation—is this within scope for our current thoughts on views, or beyond them?
- "queues": Are these memories, accessed in a special way? Do they also do synchronization?

**Comment by sa2257** [comment]
Okay. Sounds good.

- Memory and resource allocation means just binding a resource to a particular array/ function. Examples in the user guide are very ambiguous. Still trying to tease out how to use them.
- Serialization, I simply mean some way to ensure a certain block of logic happen strictly after another block. Blocking statements.
- Some data structure that will be already be populated at the time of execution. For example, to get sine and cosine, it is easier to store them than implement a function to generate them.
- I'm not entirely sure if it's within the scope, for instance to run a 3x1 filter on a 32 element array, inner `unroll` 3 should allow 32 element array access if it's banked by 3 (mismatch with 32) or larger. (we can doctor the array size for simplicity)
- I need to understand how queues get implemented in HLS. HLS allows FIFOs. Maybe it'll be a simple use of that. But I'm not clear how it gets implemented in MachSuite.

**Comment by sa2257** [comment]
| Benchmarks        | nested_loops | reductions | tiling | blocking_op | index_add | filter | while | if | bitwise | indirection | unroll | bank | pipeline | queues | LUTs | resource_bind | recursion |
|-------------------|--------------|------------|--------|-------------|-----------|--------|-------|----|---------|-------------|--------|------|----------|--------|------|---------------|-----------|
| gemm/ncubed       |       *      |      *     |    -   |      -      |     -     |        |       |    |         |             |    *   |   *  |     *    |        |      |      *        |           |
| spatial/fir       |              |            |        |      -      |     -     |        |       |    |         |             |        |      |          |        |      |               |           |
| gemm/blocked      |       *      |      *     |    *   |      -      |     -     |        |       |    |         |             |    *   |   *  |     *    |        |      | *             |           |
| stencil/stencil2d |       *      |      *     |        |      -      |     *     |    *   |       |    |         |             |    *   |   *  |     *    |        |      | *             |           |
| stencil/stencil3d |       *      |      *     |        |      -      |     *     |    *   |       |    |         |             |    *   |   *  |     *    |        |      | *             |           |
| fft/strided       |       *      |            |        |      *      |     -     |        |       |    |         |             |    *   |   *  |     *    |        |      | *             |           |
| fft/transpose     |       *      |            |        |             |     -     |        |       |    |         |             |    *   |   *  |     *    |        | *    | *             |           |
| spmv/ellpack      |       *      |            |        |             |     -     |        |       |    |         |      *      |    *   |   *  |     *    |        |      | *             |           |
| spmv/crs          |       *      |            |        |             |     *     |        |       |    |         |      *      |    *   |   *  |     *    |        |      | *             |           |
| kmp/kmp           |       *      | *          |        |             |     *     |        |   *   |  * |         |             |    *   |   *  |     *    |        |      |               |           |
| sort/merge        |       *      |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| sort/radix        |       *      |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| bfs/bulk          |       *      |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| bfs/queue         |       *      |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/kmeans    |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/gda       |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/bs        |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/pagerank  |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/sw        |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| spatial/tq6       |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| md/knn            |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| md/grid           |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| backprop/backprop |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| nw/nw             |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| aes/aes           |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |
| viterbi/viterbi   |              |            |        |             |           |        |       |    |         |             |        |      |          |        |      |               |           |

**Comment by rachitnigam** [comment]
I assume a combination of https://github.com/cucapra/fuse-benchmarks/issues/74 and direct issues are being used to track this now. @sa2257 reopen if you still need this.
